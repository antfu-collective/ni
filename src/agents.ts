const npmRun = (agent: string) => (args: string[]) => {
  if (args.length > 1)
    return `${agent} run ${args[0]} -- ${args.slice(1).join(' ')}`
  else return `${agent} run ${args[0]}`
}

export const AGENTS = {
  npm: {
    'run': npmRun('npm'),
    'install': 'npm i',
    'frozen': 'npm ci',
    'global': 'npm i -g {0}',
    'add': 'npm i {0}',
    'upgrade': 'npm update {0}',
    'upgrade-interactive': null,
    'execute': 'npx {0}',
    'uninstall': 'npm uninstall {0}',
    'global_uninstall': 'npm uninstall -g {0}',
  },
  yarn: {
    'run': 'yarn run {0}',
    'install': 'yarn install',
    'frozen': 'yarn install --frozen-lockfile',
    'global': 'yarn global add {0}',
    'add': 'yarn add {0}',
    'upgrade': 'yarn upgrade {0}',
    'upgrade-interactive': 'yarn upgrade-interactive',
    'execute': 'yarn dls {0}',
    'uninstall': 'yarn remove {0}',
    'global_uninstall': 'yarn global remove {0}',
  },
  pnpm: {
    'run': npmRun('pnpm'),
    'install': 'pnpm i',
    'frozen': 'pnpm i --frozen-lockfile',
    'global': 'pnpm i -g {0}',
    'add': 'pnpm i {0}',
    'upgrade': 'pnpm update {0}',
    'upgrade-interactive': 'pnpm update -i',
    'execute': 'pnpm dlx {0}',
    'uninstall': 'pnpm remove {0}',
    'global_uninstall': 'pnpm remove -g {0}',
  },
}

export type Agent = keyof typeof AGENTS
export type Command = keyof typeof AGENTS.npm

export const agents = Object.keys(AGENTS) as Agent[]

export const LOCKS: Record<string, Agent> = {
  'pnpm-lock.yaml': 'pnpm',
  'yarn.lock': 'yarn',
  'package-lock.json': 'npm',
}

export const INSTALL_PAGE: Record<Agent, string> = {
  pnpm: 'https://pnpm.js.org/en/installation',
  yarn: 'https://yarnpkg.com/getting-started/install',
  npm: 'https://www.npmjs.com/get-npm',
}
