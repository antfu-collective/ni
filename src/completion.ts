// Print completion script
export const rawCompletionScript = `
###-begin-nr-completion-###

if type complete &>/dev/null; then
  _nr_completion() {
    local words
    local cur
    local cword
    _get_comp_words_by_ref -n =: cur words cword
    IFS=$'\\n'
    COMPREPLY=($(COMP_CWORD=$cword COMP_LINE=$cur nr --completion \${words[@]}))
  }
  complete -F _nr_completion nr
fi

###-end-nr-completion-###
`.trim()
