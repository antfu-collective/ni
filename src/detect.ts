import process from 'node:process'
import { detect as detectPM } from 'package-manager-detector'
import { x } from 'tinyexec'
import terminalLink from 'terminal-link'
import prompts from '@posva/prompts'
import { INSTALL_PAGE } from 'package-manager-detector/constants'
import { cmdExists } from './utils'

export interface DetectOptions {
  autoInstall?: boolean
  programmatic?: boolean
  cwd?: string
  /**
   * Should use Volta when present
   *
   * @see https://volta.sh/
   * @default true
   */
  detectVolta?: boolean
}

export async function detect({ autoInstall, programmatic, cwd }: DetectOptions = {}) {
  const {
    name,
    agent,
    version,
  } = await detectPM({
    cwd,
    onUnknown: (packageManager) => {
      if (!programmatic) {
        console.warn('[ni] Unknown packageManager:', packageManager)
      }
      return undefined
    },
  }) || {}

  // auto install
  if (name && !cmdExists(name) && !programmatic) {
    if (!autoInstall) {
      console.warn(`[ni] Detected ${name} but it doesn't seem to be installed.\n`)

      if (process.env.CI)
        process.exit(1)

      const link = terminalLink(name, INSTALL_PAGE[name])
      const { tryInstall } = await prompts({
        name: 'tryInstall',
        type: 'confirm',
        message: `Would you like to globally install ${link}?`,
      })
      if (!tryInstall)
        process.exit(1)
    }

    await x(
      'npm',
      ['i', '-g', `${name}${version ? `@${version}` : ''}`],
      {
        nodeOptions: {
          stdio: 'inherit',
          cwd,
        },
        throwOnError: true,
      },
    )
  }

  return agent
}
