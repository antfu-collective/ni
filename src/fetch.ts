import process from 'node:process'
import type { Choice } from '@posva/prompts'

export interface NpmPackage {
  name: string
  description: string
  version: string
  keywords: string[]
  date: string
  links: {
    npm: string
    homepage: string
    repository: string
  }
}

interface NpmRegistryResponse {
  objects: { package: NpmPackage }[]
}

export async function fetchNpmPackages(pattern: string): Promise<Choice[]> {
  const registryLink = (pattern: string) =>
        `https://registry.npmjs.com/-/v1/search?text=${pattern}&size=35`

  try {
    const result = await fetch(registryLink(pattern))
      .then(res => res.json()) as NpmRegistryResponse

    return result.objects.map(({ package: pkg }) => ({
      title: pkg.name,
      value: pkg,
      description: `${pkg.version} - ${pkg.description}`,
    }))
  }
  catch (e) {
    console.error('Error when fetching npm registry')
    process.exit(1)
  }
}
