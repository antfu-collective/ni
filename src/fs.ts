import fs from 'node:fs'
import { resolve } from 'node:path'
import type { Agent } from './agents'

const getJSONConfigFile = (cwd: string, filename: string) => {
  const path = resolve(cwd, filename)

  if (fs.existsSync(path)) {
    try {
      const raw = fs.readFileSync(path, 'utf-8')
      const data = JSON.parse(raw) as Record<string, any>
      return data
    }
    catch (e) {
      console.warn(`Failed to parse ${filename}`)
      process.exit(0)
    }
  }
}

export function getPackageJSON(cwd = process.cwd()) {
  return getJSONConfigFile(cwd, 'package.json') || {}
}

export function getDenoJSON(cwd = process.cwd()) {
  return getJSONConfigFile(cwd, 'deno.json') || getJSONConfigFile(cwd, 'deno.jsonc') || {}
}

export function getConfig(agent: Agent, cwd = process.cwd()) {
  if (agent === 'deno')
    return getDenoJSON(cwd)
  return getPackageJSON(cwd)
}
