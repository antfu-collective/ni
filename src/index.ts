export * from 'package-manager-detector/constants'
export * from 'package-manager-detector/commands'

export * from './config'
export * from './detect'
export * from './parse'
export * from './runner'
export * from './utils'
