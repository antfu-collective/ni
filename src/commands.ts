import { Agent, AGENTS, Command } from './agents'
import { exclude } from './utils'

export function getCommand(agent: Agent, commnad: Command, args: string[] = []) {
  if (!(agent in AGENTS))
    throw new Error(`Unsupported agent "${agent}"`)

  const c = AGENTS[agent][commnad]

  if (typeof c === 'function')
    return c(args)

  return c.replace('{0}', args.join(' '))
}

export function parseNi(agent: Agent, args: string[], hasLock?: boolean): string {
  if (args.length === 0)
    return getCommand(agent, 'install')

  if (args.includes('-g'))
    return getCommand(agent, 'global', exclude(args, '-g'))

  if (args.includes('--frozen-if-present')) {
    args = exclude(args, '--frozen-if-present')
    return getCommand(agent, hasLock ? 'frozen' : 'install', args)
  }

  if (args.includes('--frozen'))
    return getCommand(agent, 'frozen', exclude(args, '--frozen'))

  return getCommand(agent, 'add', args)
}

export function parseNr(agent: Agent, args: string[]): string {
  if (args.length === 0)
    args.push('start')

  return getCommand(agent, 'run', args)
}
