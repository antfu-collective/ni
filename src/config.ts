import fs from 'node:fs'
import path from 'node:path'
import process from 'node:process'
import ini from 'ini'
import type { Agent } from 'package-manager-detector'
import { detect } from './detect'

const customRcPath = process.env.NI_CONFIG_FILE

const home = process.platform === 'win32'
  ? process.env.USERPROFILE
  : process.env.HOME

const defaultRcPath = path.join(home || '~/', '.nirc')

const rcPath = customRcPath || defaultRcPath

interface Config {
  defaultAgent: Agent | 'prompt'
  globalAgent: Agent
}

const defaultConfig: Config = {
  defaultAgent: 'prompt',
  globalAgent: 'npm',
}

let config: Config | undefined

export async function getConfig(): Promise<Config> {
  if (!config) {
    config = Object.assign(
      {},
      defaultConfig,
      fs.existsSync(rcPath)
        ? ini.parse(fs.readFileSync(rcPath, 'utf-8'))
        : null,
    )

    if (process.env.NI_DEFAULT_AGENT)
      config.defaultAgent = process.env.NI_DEFAULT_AGENT as Agent

    if (process.env.NI_GLOBAL_AGENT)
      config.globalAgent = process.env.NI_GLOBAL_AGENT as Agent

    const agent = await detect({ programmatic: true })
    if (agent)
      config.defaultAgent = agent
  }

  return config
}

export async function getDefaultAgent(programmatic?: boolean) {
  const { defaultAgent } = await getConfig()
  if (defaultAgent === 'prompt' && (programmatic || process.env.CI))
    return 'npm'
  return defaultAgent
}

export async function getGlobalAgent() {
  const { globalAgent } = await getConfig()
  return globalAgent
}
