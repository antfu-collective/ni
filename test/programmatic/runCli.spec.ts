import path from 'node:path'
import { tmpdir } from 'node:os'
import fs from 'fs-extra'
import type { MockInstance } from 'vitest'
import { afterAll, beforeAll, describe, expect, it, vi } from 'vitest'
import { AGENTS, parseNa, parseNi, parseNlx, parseNu, parseNun, runCli } from '../../src'

import type { Runner } from '../../src'

let basicLog: MockInstance, errorLog: MockInstance, warnLog: MockInstance, infoLog: MockInstance

function runCliTest(fixtureName: string, agent: string, runner: Runner, args: string[]) {
  return async () => {
    const cwd = await fs.mkdtemp(path.join(tmpdir(), 'ni-'))
    const fixture = path.join(__dirname, '..', 'fixtures', fixtureName, agent)
    await fs.copy(fixture, cwd)

    await runCli(
      async (agent, _, ctx) => {
        // we override the args to be test specific
        return runner(agent, args, ctx)
      },
      {
        programmatic: true,
        cwd,
        args,
      },
    ).catch((e) => {
      // it will always throw if ezspawn is mocked
      if (e.command)
        expect(e.command).toMatchSnapshot()
      else
        expect(e.message).toMatchSnapshot()
    })
  }
}

beforeAll(() => {
  basicLog = vi.spyOn(console, 'log')
  warnLog = vi.spyOn(console, 'warn')
  errorLog = vi.spyOn(console, 'error')
  infoLog = vi.spyOn(console, 'info')

  vi.mock('tinyexec', async (importOriginal) => {
    const mod = await importOriginal() as any
    return {
      ...mod,
      x: (cmd: string, args?: string[]) => {
        // break execution flow for easier snapshotting
        // eslint-disable-next-line no-throw-literal
        throw { command: [cmd, ...(args ?? [])].join(' ') }
      },
    }
  })
})

afterAll(() => {
  vi.resetAllMocks()
})

const agents = [...AGENTS, 'unknown']
const fixtures = ['lockfile', 'packager']

// matrix testing of: fixtures x agents x commands
fixtures.forEach(fixture => describe(fixture, () => agents.forEach(agent => describe(agent, () => {
  /** na */
  it('na', runCliTest(fixture, agent, parseNa, []))
  it('na run foo', runCliTest(fixture, agent, parseNa, ['run', 'foo']))

  /** ni */
  it('ni', runCliTest(fixture, agent, parseNi, []))
  it('ni foo', runCliTest(fixture, agent, parseNi, ['foo']))
  it('ni foo -D', runCliTest(fixture, agent, parseNi, ['foo', '-D']))
  it('ni --frozen', runCliTest(fixture, agent, parseNi, ['--frozen']))
  it('ni -g foo', runCliTest(fixture, agent, parseNi, ['-g', 'foo']))

  /** nlx */
  it('nlx', runCliTest(fixture, agent, parseNlx, ['foo']))

  /** nu */
  it('nu', runCliTest(fixture, agent, parseNu, []))
  it('nu -i', runCliTest(fixture, agent, parseNu, ['-i']))

  /** nun */
  it('nun foo', runCliTest(fixture, agent, parseNun, ['foo']))
  it('nun -g foo', runCliTest(fixture, agent, parseNun, ['-g', 'foo']))

  it('no logs', () => {
    expect(basicLog).not.toHaveBeenCalled()
    expect(warnLog).not.toHaveBeenCalled()
    expect(errorLog).not.toHaveBeenCalled()
    expect(infoLog).not.toHaveBeenCalled()
  })
}))))
