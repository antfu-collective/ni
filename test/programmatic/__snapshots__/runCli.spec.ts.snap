// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`lockfile > bun > na 1`] = `"bun"`;

exports[`lockfile > bun > na run foo 1`] = `"bun run foo"`;

exports[`lockfile > bun > ni --frozen 1`] = `"bun install --no-save"`;

exports[`lockfile > bun > ni -g foo 1`] = `"npm i -g foo"`;

exports[`lockfile > bun > ni 1`] = `"bun install"`;

exports[`lockfile > bun > ni foo -D 1`] = `"bun add foo -d"`;

exports[`lockfile > bun > ni foo 1`] = `"bun add foo"`;

exports[`lockfile > bun > nlx 1`] = `"bunx foo"`;

exports[`lockfile > bun > nu -i 1`] = `"bun update"`;

exports[`lockfile > bun > nu 1`] = `"bun update"`;

exports[`lockfile > bun > nun -g foo 1`] = `"npm uninstall -g foo"`;

exports[`lockfile > bun > nun foo 1`] = `"bun remove foo"`;

exports[`lockfile > npm > na 1`] = `"npm"`;

exports[`lockfile > npm > na run foo 1`] = `"npm run foo"`;

exports[`lockfile > npm > ni --frozen 1`] = `"npm ci"`;

exports[`lockfile > npm > ni -g foo 1`] = `"npm i -g foo"`;

exports[`lockfile > npm > ni 1`] = `"npm i"`;

exports[`lockfile > npm > ni foo -D 1`] = `"npm i foo -D"`;

exports[`lockfile > npm > ni foo 1`] = `"npm i foo"`;

exports[`lockfile > npm > nlx 1`] = `"npx foo"`;

exports[`lockfile > npm > nu -i 1`] = `"Command \\"upgrade-interactive\\" is not support by agent \\"npm\\""`;

exports[`lockfile > npm > nu 1`] = `"npm update"`;

exports[`lockfile > npm > nun -g foo 1`] = `"npm uninstall -g foo"`;

exports[`lockfile > npm > nun foo 1`] = `"npm uninstall foo"`;

exports[`lockfile > pnpm > na 1`] = `"pnpm"`;

exports[`lockfile > pnpm > na run foo 1`] = `"pnpm run foo"`;

exports[`lockfile > pnpm > ni --frozen 1`] = `"pnpm i --frozen-lockfile"`;

exports[`lockfile > pnpm > ni -g foo 1`] = `"npm i -g foo"`;

exports[`lockfile > pnpm > ni 1`] = `"pnpm i"`;

exports[`lockfile > pnpm > ni foo -D 1`] = `"pnpm add foo -D"`;

exports[`lockfile > pnpm > ni foo 1`] = `"pnpm add foo"`;

exports[`lockfile > pnpm > nlx 1`] = `"pnpm dlx foo"`;

exports[`lockfile > pnpm > nu -i 1`] = `"pnpm update -i"`;

exports[`lockfile > pnpm > nu 1`] = `"pnpm update"`;

exports[`lockfile > pnpm > nun -g foo 1`] = `"npm uninstall -g foo"`;

exports[`lockfile > pnpm > nun foo 1`] = `"pnpm remove foo"`;

exports[`lockfile > pnpm@6 > na 1`] = `"pnpm"`;

exports[`lockfile > pnpm@6 > na run foo 1`] = `"pnpm run foo"`;

exports[`lockfile > pnpm@6 > ni --frozen 1`] = `"pnpm i --frozen-lockfile"`;

exports[`lockfile > pnpm@6 > ni -g foo 1`] = `"npm i -g foo"`;

exports[`lockfile > pnpm@6 > ni 1`] = `"pnpm i"`;

exports[`lockfile > pnpm@6 > ni foo -D 1`] = `"pnpm add foo -D"`;

exports[`lockfile > pnpm@6 > ni foo 1`] = `"pnpm add foo"`;

exports[`lockfile > pnpm@6 > nlx 1`] = `"pnpm dlx foo"`;

exports[`lockfile > pnpm@6 > nu -i 1`] = `"pnpm update -i"`;

exports[`lockfile > pnpm@6 > nu 1`] = `"pnpm update"`;

exports[`lockfile > pnpm@6 > nun -g foo 1`] = `"npm uninstall -g foo"`;

exports[`lockfile > pnpm@6 > nun foo 1`] = `"pnpm remove foo"`;

exports[`lockfile > unknown > na 1`] = `"pnpm"`;

exports[`lockfile > unknown > na run foo 1`] = `"pnpm run foo"`;

exports[`lockfile > unknown > ni --frozen 1`] = `"pnpm i --frozen-lockfile"`;

exports[`lockfile > unknown > ni -g foo 1`] = `"npm i -g foo"`;

exports[`lockfile > unknown > ni 1`] = `"pnpm i"`;

exports[`lockfile > unknown > ni foo -D 1`] = `"pnpm add foo -D"`;

exports[`lockfile > unknown > ni foo 1`] = `"pnpm add foo"`;

exports[`lockfile > unknown > nlx 1`] = `"pnpm dlx foo"`;

exports[`lockfile > unknown > nu -i 1`] = `"pnpm update -i"`;

exports[`lockfile > unknown > nu 1`] = `"pnpm update"`;

exports[`lockfile > unknown > nun -g foo 1`] = `"npm uninstall -g foo"`;

exports[`lockfile > unknown > nun foo 1`] = `"pnpm remove foo"`;

exports[`lockfile > yarn > na 1`] = `"yarn"`;

exports[`lockfile > yarn > na run foo 1`] = `"yarn run foo"`;

exports[`lockfile > yarn > ni --frozen 1`] = `"yarn install --frozen-lockfile"`;

exports[`lockfile > yarn > ni -g foo 1`] = `"npm i -g foo"`;

exports[`lockfile > yarn > ni 1`] = `"yarn install"`;

exports[`lockfile > yarn > ni foo -D 1`] = `"yarn add foo -D"`;

exports[`lockfile > yarn > ni foo 1`] = `"yarn add foo"`;

exports[`lockfile > yarn > nlx 1`] = `"npx foo"`;

exports[`lockfile > yarn > nu -i 1`] = `"yarn upgrade-interactive"`;

exports[`lockfile > yarn > nu 1`] = `"yarn upgrade"`;

exports[`lockfile > yarn > nun -g foo 1`] = `"npm uninstall -g foo"`;

exports[`lockfile > yarn > nun foo 1`] = `"yarn remove foo"`;

exports[`lockfile > yarn@berry > na 1`] = `"yarn"`;

exports[`lockfile > yarn@berry > na run foo 1`] = `"yarn run foo"`;

exports[`lockfile > yarn@berry > ni --frozen 1`] = `"yarn install --frozen-lockfile"`;

exports[`lockfile > yarn@berry > ni -g foo 1`] = `"npm i -g foo"`;

exports[`lockfile > yarn@berry > ni 1`] = `"yarn install"`;

exports[`lockfile > yarn@berry > ni foo -D 1`] = `"yarn add foo -D"`;

exports[`lockfile > yarn@berry > ni foo 1`] = `"yarn add foo"`;

exports[`lockfile > yarn@berry > nlx 1`] = `"npx foo"`;

exports[`lockfile > yarn@berry > nu -i 1`] = `"yarn upgrade-interactive"`;

exports[`lockfile > yarn@berry > nu 1`] = `"yarn upgrade"`;

exports[`lockfile > yarn@berry > nun -g foo 1`] = `"npm uninstall -g foo"`;

exports[`lockfile > yarn@berry > nun foo 1`] = `"yarn remove foo"`;

exports[`packager > bun > na 1`] = `"bun"`;

exports[`packager > bun > na run foo 1`] = `"bun run foo"`;

exports[`packager > bun > ni --frozen 1`] = `"bun install --no-save"`;

exports[`packager > bun > ni -g foo 1`] = `"npm i -g foo"`;

exports[`packager > bun > ni 1`] = `"bun install"`;

exports[`packager > bun > ni foo -D 1`] = `"bun add foo -d"`;

exports[`packager > bun > ni foo 1`] = `"bun add foo"`;

exports[`packager > bun > nlx 1`] = `"bunx foo"`;

exports[`packager > bun > nu -i 1`] = `"bun update"`;

exports[`packager > bun > nu 1`] = `"bun update"`;

exports[`packager > bun > nun -g foo 1`] = `"npm uninstall -g foo"`;

exports[`packager > bun > nun foo 1`] = `"bun remove foo"`;

exports[`packager > npm > na 1`] = `"npm"`;

exports[`packager > npm > na run foo 1`] = `"npm run foo"`;

exports[`packager > npm > ni --frozen 1`] = `"npm ci"`;

exports[`packager > npm > ni -g foo 1`] = `"npm i -g foo"`;

exports[`packager > npm > ni 1`] = `"npm i"`;

exports[`packager > npm > ni foo -D 1`] = `"npm i foo -D"`;

exports[`packager > npm > ni foo 1`] = `"npm i foo"`;

exports[`packager > npm > nlx 1`] = `"npx foo"`;

exports[`packager > npm > nu -i 1`] = `"Command \\"upgrade-interactive\\" is not support by agent \\"npm\\""`;

exports[`packager > npm > nu 1`] = `"npm update"`;

exports[`packager > npm > nun -g foo 1`] = `"npm uninstall -g foo"`;

exports[`packager > npm > nun foo 1`] = `"npm uninstall foo"`;

exports[`packager > pnpm > na 1`] = `"pnpm"`;

exports[`packager > pnpm > na run foo 1`] = `"pnpm run foo"`;

exports[`packager > pnpm > ni --frozen 1`] = `"pnpm i --frozen-lockfile"`;

exports[`packager > pnpm > ni -g foo 1`] = `"npm i -g foo"`;

exports[`packager > pnpm > ni 1`] = `"pnpm i"`;

exports[`packager > pnpm > ni foo -D 1`] = `"pnpm add foo -D"`;

exports[`packager > pnpm > ni foo 1`] = `"pnpm add foo"`;

exports[`packager > pnpm > nlx 1`] = `"pnpm dlx foo"`;

exports[`packager > pnpm > nu -i 1`] = `"pnpm update -i"`;

exports[`packager > pnpm > nu 1`] = `"pnpm update"`;

exports[`packager > pnpm > nun -g foo 1`] = `"npm uninstall -g foo"`;

exports[`packager > pnpm > nun foo 1`] = `"pnpm remove foo"`;

exports[`packager > pnpm@6 > na 1`] = `"pnpm"`;

exports[`packager > pnpm@6 > na run foo 1`] = `"pnpm run foo"`;

exports[`packager > pnpm@6 > ni --frozen 1`] = `"pnpm i --frozen-lockfile"`;

exports[`packager > pnpm@6 > ni -g foo 1`] = `"npm i -g foo"`;

exports[`packager > pnpm@6 > ni 1`] = `"pnpm i"`;

exports[`packager > pnpm@6 > ni foo -D 1`] = `"pnpm add foo -D"`;

exports[`packager > pnpm@6 > ni foo 1`] = `"pnpm add foo"`;

exports[`packager > pnpm@6 > nlx 1`] = `"pnpm dlx foo"`;

exports[`packager > pnpm@6 > nu -i 1`] = `"pnpm update -i"`;

exports[`packager > pnpm@6 > nu 1`] = `"pnpm update"`;

exports[`packager > pnpm@6 > nun -g foo 1`] = `"npm uninstall -g foo"`;

exports[`packager > pnpm@6 > nun foo 1`] = `"pnpm remove foo"`;

exports[`packager > unknown > na 1`] = `"pnpm"`;

exports[`packager > unknown > na run foo 1`] = `"pnpm run foo"`;

exports[`packager > unknown > ni --frozen 1`] = `"pnpm i --frozen-lockfile"`;

exports[`packager > unknown > ni -g foo 1`] = `"npm i -g foo"`;

exports[`packager > unknown > ni 1`] = `"pnpm i"`;

exports[`packager > unknown > ni foo -D 1`] = `"pnpm add foo -D"`;

exports[`packager > unknown > ni foo 1`] = `"pnpm add foo"`;

exports[`packager > unknown > nlx 1`] = `"pnpm dlx foo"`;

exports[`packager > unknown > nu -i 1`] = `"pnpm update -i"`;

exports[`packager > unknown > nu 1`] = `"pnpm update"`;

exports[`packager > unknown > nun -g foo 1`] = `"npm uninstall -g foo"`;

exports[`packager > unknown > nun foo 1`] = `"pnpm remove foo"`;

exports[`packager > yarn > na 1`] = `"yarn"`;

exports[`packager > yarn > na run foo 1`] = `"yarn run foo"`;

exports[`packager > yarn > ni --frozen 1`] = `"yarn install --frozen-lockfile"`;

exports[`packager > yarn > ni -g foo 1`] = `"npm i -g foo"`;

exports[`packager > yarn > ni 1`] = `"yarn install"`;

exports[`packager > yarn > ni foo -D 1`] = `"yarn add foo -D"`;

exports[`packager > yarn > ni foo 1`] = `"yarn add foo"`;

exports[`packager > yarn > nlx 1`] = `"npx foo"`;

exports[`packager > yarn > nu -i 1`] = `"yarn upgrade-interactive"`;

exports[`packager > yarn > nu 1`] = `"yarn upgrade"`;

exports[`packager > yarn > nun -g foo 1`] = `"npm uninstall -g foo"`;

exports[`packager > yarn > nun foo 1`] = `"yarn remove foo"`;

exports[`packager > yarn@berry > na 1`] = `"yarn"`;

exports[`packager > yarn@berry > na run foo 1`] = `"yarn run foo"`;

exports[`packager > yarn@berry > ni --frozen 1`] = `"yarn install --immutable"`;

exports[`packager > yarn@berry > ni -g foo 1`] = `"npm i -g foo"`;

exports[`packager > yarn@berry > ni 1`] = `"yarn install"`;

exports[`packager > yarn@berry > ni foo -D 1`] = `"yarn add foo -D"`;

exports[`packager > yarn@berry > ni foo 1`] = `"yarn add foo"`;

exports[`packager > yarn@berry > nlx 1`] = `"yarn dlx foo"`;

exports[`packager > yarn@berry > nu -i 1`] = `"yarn up -i"`;

exports[`packager > yarn@berry > nu 1`] = `"yarn up"`;

exports[`packager > yarn@berry > nun -g foo 1`] = `"npm uninstall -g foo"`;

exports[`packager > yarn@berry > nun foo 1`] = `"yarn remove foo"`;
